ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,545 @@
+initRangeReadLMDBTable :: LedgerConstraint k v => Int -> LMDBMK k v                 -> LMDB.Transaction mode           (ValuesMK k v)
+rangeReadLMDBTable     :: LedgerConstraint k v => Int -> LMDBMK k v -> KeysMK   k v -> LMDB.Transaction mode           (ValuesMK k v)
+
+getDb                                                  (NameMK   name) = LMDBMK name <$> LMDB.getDatabase (Just name)
+initLMDBTable                (LMDBMK tbl_name db) (ApplyValuesMK vals) = ApplyEmptyMK     <$  lmdbInitTable  tbl_name db vals
+writeLMDBTable               (LMDBMK _        db) (ApplyDiffMK   diff) = ApplyEmptyMK     <$  lmdbWriteTable          db diff
+readLMDBTable                (LMDBMK _        db) (ApplyKeysMK   keys) = ApplyValuesMK    <$> lmdbReadTable           db keys
+initRangeReadLMDBTable count (LMDBMK _        db)                      = ApplyValuesMK    <$> lmdbInitRangeReadTable count db
+rangeReadLMDBTable     count (LMDBMK _        db) (ApplyKeysMK   keys) = ApplyValuesMK    <$> lmdbRangeReadTable count db keys
+
+{-------------------------------------------------------------------------------
+ Db settings
+-------------------------------------------------------------------------------}
+
+readDbSettings ::
+     LMDB.Database () DbState
+  -> LMDB.Transaction mode DbState
+readDbSettings db = readDbSettingsMaybeNull db >>= \case
+  Just x -> pure x
+  -- FIXME: if this (and other errors) are thrown, the async thread will die but
+  -- the main thread will be waiting on a TMVar, thus deadlocking the program.
+  --
+  -- This has to be fixed, properly handled.
+  Nothing -> Exn.throw DbErrNoSettings
+
+readDbSettingsMaybeNull ::
+     LMDB.Database () DbState
+  -> LMDB.Transaction mode (Maybe DbState)
+readDbSettingsMaybeNull db = LMDB.get db ()
+
+withDbSettingsRW ::
+     LMDB.Database () DbState
+  -> (DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState))
+  -> LMDB.Transaction LMDB.ReadWrite a
+withDbSettingsRW db f = withDbSettingsRWMaybeNull db $ maybe (Exn.throw DbErrNoSettings) f
+
+withDbSettingsRWMaybeNull ::
+      LMDB.Database () DbState
+   -> (Maybe DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState))
+   -> LMDB.Transaction LMDB.ReadWrite a
+withDbSettingsRWMaybeNull db f  =
+  readDbSettingsMaybeNull db >>= f >>= \(r, new_s) -> LMDB.put db () (Just new_s) $> r
+
+data GuardDbDir  = GDDMustExist | GDDMustNotExist
+
+guardDbDir :: MonadIO m => GuardDbDir -> FS.SomeHasFS m -> FS.FsPath -> m FilePath
+guardDbDir gdd (FS.SomeHasFS fs) path = do
+  FS.doesFileExist fs path >>= \b -> when b $ Exn.throw $ DbErrStr $ "guardDbDir:must be a directory:" <> show path
+  FS.doesDirectoryExist fs path >>= \case
+    True | GDDMustNotExist <- gdd ->
+           -- TODO Should throw a DbErrDbExists exception
+           -- Callers should delete the directory if they want to restore a snapshot
+           FS.removeDirectoryRecursive fs path
+    False | GDDMustExist <- gdd -> Exn.throw $ DbErrDbDoesntExist path
```

Comment:
This `throw`s exceptions, but probably the function should instead be `-> ExceptT DBErr m FilePath` to be able to act on the errors instead of needing to be in IO to catch them.